### PR TITLE
refactoring

### LINKED ISSUE #28: No config.js in project file
Update needed

### PR DESCRIPTION
**Pull request by derisen**
# CHANGELOG

* Separated authentication with popup vs. redirect flows into different files.
* Added fallback check and a backup CDN for importing msal.js
* Moved loginRequest and tokenRequest objects into authConfig.js
* Minor fixes on Readme.md

### CODE REVIEW

**Comment by derisen** [comment]
@archieag FYI

File: JavaScriptSPA/index.html [hunk 1]
```diff
@@ -6,24 +6,29 @@
...
     
     <!-- IE support: add promises polyfill before msal.js  -->    
-    <script type="text/javascript" src="//cdn.jsdelivr.net/npm/bluebird@3.7.2/js/browser/bluebird.min.js"></script>
-    <script type="text/javascript" src="https://alcdn.msftauth.net/lib/1.2.1/js/msal.js" integrity="sha384-9TV1245fz+BaI+VvCjMYL0YDMElLBwNS84v3mY57pXNOt6xcUYch2QLImaTahcOP" crossorigin="anonymous"></script>
+    <!-- polyfilling promise -->
+    <script type="text/javascript" src="https://cdn.jsdelivr.net/npm/es6-promise/dist/es6-promise.auto.min.js"></script>
+    <!-- polyfilling fetch -->
+    <script type="text/javascript" src="https://cdn.jsdelivr.net/npm/whatwg-fetch/dist/fetch.umd.min.js"></script>
+    <!-- depending on your browser you might want to include babel polyfill -->
+    <script type="text/javascript" src="https://cdn.jsdelivr.net/npm/@babel/polyfill@7.4.4/dist/polyfill.min.js"></script>
```
Comment (jasonnutter) [inline JavaScriptSPA/index.html]: Why is this needed?

File: README.md [hunk 2]
```diff
@@ -4,7 +4,6 @@ languages:
 - javascript
 products:
 - microsoft-identity-platform
-- azure-active-directory-v2
```
Comment (jasonnutter) [inline README.md]: Still seems relevant, why was it removed?

File: JavaScriptSPA/index.html [hunk 3]
```diff
@@ -62,11 +67,12 @@ <h5 class="card-title" id="WelcomeMessage">Please sign-in to see your profile an
...
     <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.4.1/js/bootstrap.min.js" integrity="sha384-wfSDF2E50Y2D1uUdj0O3uMBJnjuUD4Ih7YwaYd1iqfktj0Uod8GCExl3Og8ifwB6" crossorigin="anonymous"></script>
     
-    <!-- importing app scripts | load order is important -->
+    <!-- importing app scripts (load order is important) -->
     <script type="text/javascript" src="./authConfig.js"></script>
     <script type="text/javascript" src="./graphConfig.js"></script>
-    <script type="text/javascript" src="./ui.js"></script>  
-    <script type="text/javascript" src="./auth.js"></script>
+    <script type="text/javascript" src="./ui.js"></script>
+    <!-- replace next with authRedirect.js if you would like to use redirect flow -->
```
Comment (jasonnutter) [inline JavaScriptSPA/index.html]: If you do this, have the script tag already there (commented out).

File: JavaScriptSPA/authPopup.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+function getTokenPopup(request) {
+  return myMSALObj.acquireTokenSilent(request)
+    .catch(error => {
+      console.log("silent token acquisition fails. acquiring token using popup");
+          
+      // fallback to interaction when silent call fails
+        return myMSALObj.acquireTokenPopup(request)
+          .then(tokenResponse => {})
```
Comment (jasonnutter) [inline JavaScriptSPA/authPopup.js]: This will result in nothing getting returned from `getTokenPopup`. Either omit this `.then`, or make sure this chain returns `tokenResponse`.

File: JavaScriptSPA/authPopup.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        showWelcomeMessage(myMSALObj.getAccount());
+      }
+    }).catch((error) => {
+      console.log(error);
+    });
+}
+
+function signOut() {
+  myMSALObj.logout();
+}
```
Comment (jasonnutter) [inline JavaScriptSPA/authPopup.js:21]: Anything like this that is the same across both files should be factored into a single place, instead of being duplicated, in my opinion.

File: CHANGELOG.md [hunk 6]
```diff
@@ -7,4 +7,9 @@
 * Added new feature for reading user mails through MS Graph API.
 * UI redesigned with Bootstrap 4 framework.
 * Added AppCreationScripts for Powershell users.
-* Added tokenization for for sovereign clouds and graph endpoints. 
+* Added tokenization for for sovereign clouds and graph endpoints.
+
+## 3/09/2020
+
+* Separated authentication with popup vs. redirect flows into different files.
+* Further refactoring changes.
```
Comment (jasonnutter) [inline CHANGELOG.md]: Put the new entries at the top (reverse chronological)

File: JavaScriptSPA/index.html [hunk 7]
```diff
@@ -6,24 +6,29 @@
...
+    <script type="text/javascript" src="https://cdn.jsdelivr.net/npm/es6-promise/dist/es6-promise.auto.min.js"></script>
+    <!-- polyfilling fetch -->
+    <script type="text/javascript" src="https://cdn.jsdelivr.net/npm/whatwg-fetch/dist/fetch.umd.min.js"></script>
+    <!-- depending on your browser you might want to include babel polyfill -->
+    <script type="text/javascript" src="https://cdn.jsdelivr.net/npm/@babel/polyfill@7.4.4/dist/polyfill.min.js"></script>
     
+        <!-- msal.js with a fallback to backup CDN -->
+    <script type="text/javascript" src="https://alcdn.msauth.net/lib/1.2.1/js/msal.js" integrity="sha384-9TV1245fz+BaI+VvCjMYL0YDMElLBwNS84v3mY57pXNOt6xcUYch2QLImaTahcOP" crossorigin="anonymous"></script>
+    <script type="text/javascript">
+      if(typeof Msal === 'undefined')document.write(unescape("%3Cscript src='https://alcdn.msftauth.net/lib/1.1.3/js/msal.js' type='text/javascript' integrity='sha384-m/3NDUcz4krpIIiHgpeO0O8uxSghb+lfBTngquAo2Zuy2fEF+YgFeP08PWFo5FiJ' crossorigin='anonymous'%3E%3C/script%3E"));
```
Comment (jasonnutter) [inline JavaScriptSPA/index.html]: This is referencing 1.1.3, instead of 1.2.1.

File: JavaScriptSPA/index.html [hunk 8]
```diff
@@ -6,24 +6,29 @@
...
-          <button type="button" id="SignIn" class="btn btn-secondary dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
-            Sign In
-          </button>
-          <div class="dropdown-menu">
-            <button class="dropdown-item" id="Popup" onclick="signIn(this.id)">Sign in using Popup</button>
-            <button class="dropdown-item" id="Redirect" onclick="signIn(this.id)">Sign in using Redirect</button>
-          </div>
-      </div>
+        <button type="button" id="signIn" class="btn btn-secondary" onclick="signIn()">Sign In</button>
+        <button type="button" id="signOut" class="btn btn-success" onclick="signOut()" style="display:none">Sign Out</button>
```
Comment (jasonnutter) [inline JavaScriptSPA/index.html]: You can use the `d-none` class from bootstrap instead of inline css: https://getbootstrap.com/docs/4.4/utilities/display/#hiding-elements

File: JavaScriptSPA/ui.js [hunk 9]
```diff
@@ -11,14 +12,13 @@ function showWelcomeMessage(account) {
     // Reconfiguring DOM elements
     cardDiv.style.display = 'initial';
     welcomeDiv.innerHTML = `Welcome ${account.name}`;
-    signInButton.nextElementSibling.style.display = 'none';
-    signInButton.setAttribute("onclick", "signOut();");
-    signInButton.setAttribute('class', "btn btn-success")
-    signInButton.innerHTML = "Sign Out";
+    signInButton.style.display = 'none';
+    signOutButton.style.display = 'initial';
```
Comment (jasonnutter) [inline JavaScriptSPA/ui.js]: Instead of manipulating `style`, add/remove classes (in this case, `d-none`) from `signInButton.classList`

File: JavaScriptSPA/ui.js [hunk 10]
```diff
@@ -11,14 +12,13 @@ function showWelcomeMessage(account) {
...
 }
 
 function updateUI(data, endpoint) {
   console.log('Graph API responded at: ' + new Date().toString());
+  console.log(data)
```
Comment (jasonnutter) [inline JavaScriptSPA/ui.js]: I would say any time you console log something in this sample, add a label so its clear what the data is.

**Review by jasonnutter — CHANGES_REQUESTED** [review]

File: README.md (hunk 2 above)
Comment (derisen) [reply README.md]: I had the impression that for new samples `microsoft identity platform` should replace `azure active directory v2` (at least that's what dotnet samples are doing)

File: JavaScriptSPA/ui.js (hunk 10 above)
Comment (derisen) [reply JavaScriptSPA/ui.js]: My bad, I forgot this. Removing...

File: JavaScriptSPA/index.html (hunk 7 above)
Comment (derisen) [reply JavaScriptSPA/index.html]: right, so sorry : )

File: JavaScriptSPA/authPopup.js (hunk 5 above)
Comment (derisen) [reply JavaScriptSPA/authPopup.js:21]: I know you have a valid point but I couldn't figure out where else this would fit logically. To keep the convenience of just switching authPopup with authRedirect, this part above should either go to a new js file or to the index.html, neither is appealing... @sameerag

File: JavaScriptSPA/authPopup.js (hunk 4 above)
Comment (derisen) [reply JavaScriptSPA/authPopup.js]: I see. kept it there to keep the pattern:

```javascript
        return myMSALObj.acquireTokenPopup(request)
          .then(tokenResponse => {
            return tokenResponse;
          }).catch(error => {
            console.log(error);
          });
```

File: JavaScriptSPA/index.html (hunk 1 above)
Comment (derisen) [reply JavaScriptSPA/index.html]: I was adding pollyfills for ie11 support but there are a bunch of other things that still don't work and forcing a very restricted js usage (at least without something like npm and babel). 

Happily though, I've just been told by a few CxP people that aad samples can drop support for ie11. Removing this part then.

**Review by pkanher617 — APPROVED** [review]
This looks fine to me, I was able to run the redirect and popup sample as expected.

**Comment by derisen** [comment]
Thank you very much for all the reviews! Merging this now..
